docs/design/2023-09-21-tiflash-resource-control.md

```diff
@@ -0,0 +1,72 @@
+2. **Pipeline task:** The smallest execution unit in the Pipeline execution model.
+3. **TaskScheduler:** Responsible for scheduling Pipeline tasks.
+4. **TaskQueue:** Responsible for storing all Pipeline tasks in a queue, from which TaskScheduler retrieves tasks and executes them.
+5. **LocalAdmissionController (LAC):** As part of the distributed token bucket algorithm, it retrieves tokens from GAC and performs flow control.
+6. **GlobalAdmissionController (GAC):** As the central node of the distributed token bucket algorithm, it controls token distribution to various LACs.
+7. **Resource group:** A [resource group](https://github.com/pingcap/tidb/blob/master/docs/design/2022-11-25-global-resource-control.md#detailed-design) corresponds to a set of business collections, and each LAC has its token bucket associated with it.
+
+### Overall Architecture
+The overall design consists of the following two parts:
+
+1. **ResourceControlQueue:** This includes the main flow control and priority scheduling logic.
+2. **LocalAdmissionController:** Used for communication with [GAC](https://github.com/pingcap/tidb/blob/master/docs/design/2022-11-25-global-resource-control.md#global-quota-control--global-admission-control) to obtain tokens.
+
+#### ResourceControlQueue
+##### Task Retrieval
+To support resource control, a new type of TaskQueue, called ResourceControlQueue, has been introduced. The core logic for flow control and priority scheduling is encapsulated in this TaskQueue.
+
+The core data structure in ResourceControlQueue is a priority queue, where each element represents a ResourceGroupInfo. This priority queue is sorted by the priority of resource groups.
+
+When TaskScheduler retrieves a task to execute from the ResourceControlQueue, it gives priority to resource groups with higher priorities. Additionally, to achieve flow control, if the tokens for the corresponding resource group have been depleted, it will wait until tokens are refilled.
+
+ResourceControlQueue uses a MultiLevelFeedbackQueue to store all pipeline tasks for a specific resource group. After obtaining the MultiLevelFeedbackQueue corresponding to a specific resource group based on priority strategy, calling its take() method can fetch the specific task.
+
+##### Updating Resource Usage
+After TaskScheduler obtains a task, it will execute for 100ms and update its CPU usage status. During this time, it will call relevant interfaces of the LocalAdmissionController (LAC) to update token consumption.
+
+### LocalAdmissionController (LAC)
+LAC is responsible for managing the metadata of all resource groups on a TiFlash node, including:
+1. Recording the priorities, configurations, and TokenBuckets of all currently known resource groups.
+2. Watching GAC etcd, updating resource group configurations and detecting resource group deletions.
```

Comment:
why does it detect deletions only? Does creation be support automatically?

Reply:
When the first query of specific resource comes to TiFlash, TiFlash will create it. Etcd only take cares of deletion and config update of resource group .

Reply:
Maybe we can describe the process of "new resource group creation" somewhere. Just a personal suggestion.

Reply:
done